File: ctci/tmp/chapter_3_stacks_and_queues/python/queue.py
# First in, first out.
class Queue():
    def __init__(self):
        """
        Queue constructor.
        """
        self.queue = []
        self.length = 0

    def enqueue(self, value: int):
        """
        Adds an element to the front of a queue.
        """
        self.queue.append(value)
        self.length += 1

    def dequeue(self) -> int:
        """
        Pops an element off the front of a queue.
        """
        self.length -= 1
        return self.queue.pop(0)

    def peek(self) -> int:
        """
        Peeks at the element at the front of the queue.
        """
        return self.queue[0]

    def is_empty(self) -> bool:
        """
        Determines if a queue is empty.
        """
        return self.length == 0


if __name__ == 'main':
    q = Queue()
    for i in range(5):
        q.enqueue(i)
    print(q.queue)  # [0, 1, 2, 3, 4]
    print(q.dequeue())  # 0
    print(q.peek())  # 1

File: ctci/tmp/chapter_3_stacks_and_queues/js/queue.js
/* First in, first out. */
class Queue {
  /**
   * Queue constructor.
   */
  constructor() {
    this.queue = []
    this.length = 0
  }

  /**
   * Adds an element to the front of a queue.
   * @param   {number}  value
   */
  enqueue(value) {
    this.queue.push(value)
    this.length++
  }

  /**
   * Pops an element off the front of a queue.
   * @returns {number}
   */
  dequeue() {
    this.length--
    return this.queue.shift()
  }

  /**
   * Determines if a queue is empty.
   * @returns {boolean}
   */
  isEmpty() {
    return this.length == 0
  }

  /**
   * Peeks at the element at the front of the queue.
   * @returns {number}
   */
  peek() {
    return this.queue[0]
  }
}

module.exports = {
  Queue
}

const q = new Queue()
const arr = [4, 5, 2, 8, 1, 9, 6, 7, 3, 0]
for (let i of arr) q.enqueue(i)
console.log(q)
/*
Queue { queue: [ 4, 5, 2, 8, 1, 9, 6, 7, 3, 0 ], length: 10 }
*/
console.log(q.dequeue()) // 4
console.log(q)
/*
Queue { queue: [ 5, 2, 8, 1, 9, 6, 7, 3, 0 ], length: 9 } 
*/
console.log(q.peek()) // 5
